Add basic OpenGraph support via a partial

MattSLangford opened the pull request:
In the [conversation for common head elements](https://github.com/microdotblog/theme-blank/pull/2), adding basic OpenGraph support was discussed. 

I created a starting point with this pull request. I'm sure it'll need tweaking as there are many different possibilities. The initial goal is to have an easy default way to include OpenGraph metadata that will work for the most amount of sites. Theme Devs or Tweakers that want more control would need to handle that themselves.

### Weakness(es)

1. Right now, I think the biggest weakness with including OpenGraph has to do with images. As it is now, the home page uses the user's avatar as the image. Posts and pages will use included photos. But what if they don't have included photos? I'm guessing it should fallback to the avatar in this case? If that's the case, what's the best way to accomplish that.
2. Image sizes. Some apps parse the open graph data looking for a specific image size (1200x630). When an image isn't that size, some just skip over it while others can still use it elegantly.

### My Initial Decisions

For the open graph titles of title-less posts, I went with the blog title itself. I debated going with a very truncated summary, but I think that opens up a few more wormholes. If a developer wanted to go that route, perhaps that should be done on a per theme basis?

### Further Thoughts

I'm approaching this from my specific use case, and I know there are many other ways to go with this.

MattSLangford committed:
Create microblog_opengraph

Adding a partial to include basic open graph info.

MattSLangford committed:
Rename microblog_opengraph to microblog_opengraph.html

MattSLangford committed:
Add opengraph partial to head

manton commented:
Cool, thank you! I don't think I have strong opinions about the title and images, but let's keep this open for a few days in case other people want to chime in with their thoughts.

manton commented:
For completeness, it's probably worth pointing out that this will now make OpenGraph the default for all blogs and obsolete a couple plug-ins. Based on feedback I usually hear, I think this is what people want, so I'm good with it.

svendahlstrand commented:
I think it's super nice to suggest including Open Graph and Twitter Cards (that's still what they are called, apparently) to theme developers like this. 👏

I have experience implementing both, but not in a Micro.blog or Hugo context. What would you say are the main differences and advantages of your implementation compared to Hugo's built-in templates? ([twitter_cards.html](https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/twitter_cards.html) and [opengraph.html](https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/opengraph.html))

Also, if you want to trim away excessive HTML, when the content of the Twitter Card and Open Graph metadata is identical, you can remove the Twitter version. [From their documentation](https://developer.twitter.com/en/docs/twitter-for-websites/cards/guides/getting-started#opengraph):

> When the Twitter card processor looks for tags on a page, it first checks for the Twitter-specific property, and if not present, falls back to the supported Open Graph property. This allows for both to be defined on the page independently, and minimizes the amount of duplicate markup required to describe content and experience.

MattSLangford commented:
@svendahlstrand There was some confusion a few weeks ago about whether Twitter would still be playing nice with Opengraph. But I can't seem to find that article/thread at the moment. Maybe that's been resolved. In either case, it probably is best to lean into OG and not be overly concerned with what Twitter may or may not do.

As for Hugo's built-in templates, I think the failing was posts-without-titles and Micro.blog descriptions. Looking at it now, it might be best to build off of it instead of what some of the other themes (including Tiny) have done to date. It really might be as simple as taking the default, modifying it to accommodate title-less posts and Micro.blog descriptions, and moving on. That would be fantastic.

jsonbecker commented on layouts/partials/microblog_opengraph.html:
```diff
@@ -0,0 +1,26 @@
+{{ end }}
+
+{{ range .Params.photos }}
+  <meta name="twitter:image" content="{{ . }}">
+  <meta property="og:image" content="{{ . }}">
```
You could use `"https://micro.blog/photos/1200x/{{ . }}"` to get an image that is resized to 1200x and maintains aspect ratio for the height to address the expected dimensions.

MattSLangford replied:
Ahh, I forgot Micro.blog had that built in! Good call.

jsonbecker commented:
> @svendahlstrand There was some confusion a few weeks ago about whether Twitter would still be playing nice with Opengraph. But I can't seem to find that article/thread at the moment. Maybe that's been resolved. In either case, it probably is best to lean into OG and not be overly concerned with what Twitter may or may not do.
> 
> As for Hugo's built-in templates, I think the failing was posts-without-titles and Micro.blog descriptions. Looking at it now, it might be best to build off of it instead of what some of the other themes (including Tiny) have done to date. It really might be as simple as taking the default, modifying it to accommodate title-less posts and Micro.blog descriptions, and moving on. That would be fantastic.

FWIW, I've used the internal template and not seen that failure, but then again, only titled posts are routinely shared with links so maybe I'm wrong.

Also, my experience even before the Musk takeover is that getting Twitter to play nice was always easier by doubling up-- it very often did not successfully "fall back" in ways that were unpredictable. So in my opinion, having both for coverage is appropriate.

MattSLangford commented:
@jsonbecker - Good notes on that. I'll probably setup a test blog with the pure defaults and do some testing. If the weak points around title-less posts and descriptions are there, I'll try to remedy that. Don't think it'd be too hard either way.

MattSLangford committed:
Rewritten with Hugo's default as a base model

MattSLangford commented:
Alright, lots to detail now. 

@svendahlstrand I used Hugo's built-in templates as a base, but I modified it for Micro.blog specific use cases.

@jsonbecker I'm using both Twitter and OG.

### Here are the changes

**First**, I combined Hugo's OG template and Twitter template into a single file.

**Next**, I tested the 6 main types of posts and pages one might encounter on a typical Micro.blog site. These are:

- Home Page
- Traditional Page
- Post with Title (with Image)
- Post with Title (without Image)
- Post without Title (with Image)
- Post without Title (without Image)

Here are the things that didn't work with the default Hugo implementation and my fixes:

- **Home Page**: No image would display. I changed this to include the users Micro.blog avatar as the image. It would also only show "Follow on Micro.blog" as the description. I changed that to show the user's Micro.blog *about* section.
- **Traditional Page**: Added avatar as a backup image if no other images are present.
- **Single Posts (all versions)**: Added avatar as backup image if no other images are present.
- **Single Posts with No Title**: Added Site Title as Title

I also had to make a few other small changes such as:

- Properly displaying user's Twitter handle
- Removing code (such as featured images) that doesn't deal with Micro.blog

Lastly, I tested every post/page variation at both opengraph.xyz and https://iplocation.io/open-graph-checker.

Have a look and let me know if you have any further thoughts. 

@manton - I think this covers the use-cases of 99% of Micro.blog users (and web users in general). I'm sure there are a few one-offs that have different needs, so I think it was smart (as you suggested) to break this out from the microblog_head partial.

mandaris commented:
I like it. The only thing I would change is removing the last line about twitter because that is found in the microblog_head.html.

Sorry I'm so late to respond.

MattSLangford committed:
Update microblog_opengraph.html

Removed duplicate lines

MattSLangford commented:
> I like it. The only thing I would change is removing the last line about twitter because that is found in the microblog_head.html.
> 
> Sorry I'm so late to respond.

For some reason, just saw this. Good catch. I removed that.
